src/network/networkd-wwan-bus.c

```diff
@@ -0,0 +1,364 @@
+                        "GetAll",
+                        bearer_get_all_handler,
+                        b,
+                        NULL);
+        if (r < 0)
+                return log_warning_errno(r, "Could not get properties of bearer \"%s\": %m", b->path);
+
+        return 0;
+}
+
+static int bearer_new_and_initialize(Manager *manager, const char *path) {
+        _cleanup_(bearer_freep) Bearer *b = NULL;
+        int r;
+
+        assert(manager);
+        assert(path);
+
+        r = bearer_new(manager, path, &b);
+        if (r < 0)
+                return log_warning_errno(r, "Failed to allocate new bearer \"%s\": %m", path);
+
+        r = bearer_initialize(b);
+        if (r < 0)
+                return r;
+
+        TAKE_PTR(b);
+        return 0;
+}
+
+static int bearer_save_path(const char *path, void *userdata) {
+        Set **set = ASSERT_PTR(userdata);
+
+        return set_put_strdup(set, path);
+}
+
+static int enumerate_bearer_handler(sd_bus_message *message, void *userdata, sd_bus_error *ret_error) {
+        static const XMLIntrospectOps ops = {
+                .on_path = bearer_save_path,
+        };
+
+        Manager *manager = ASSERT_PTR(userdata);
+        _cleanup_set_free_ Set *paths = NULL;
+        const sd_bus_error *e;
+        const char *xml, *path;
+        int r;
+
+        assert(message);
+
+        e = sd_bus_message_get_error(message);
+        if (e) {
+                int level = LOG_WARNING;
+
+                if (sd_bus_error_has_name(e, SD_BUS_ERROR_SERVICE_UNKNOWN))
+                        /* ModemManager is not started yet. */
+                        level = LOG_DEBUG;
+
+                r = sd_bus_error_get_errno(e);
+                log_full_errno(level, r, "Could not get bearers: %s", bus_error_message(e, r));
+                return 0;
+        }
+
+        r = sd_bus_message_read(message, "s", &xml);
+        if (r < 0)
+                return bus_log_parse_error(r);
+
+        r = parse_xml_introspect("/org/freedesktop/ModemManager/Bearer", xml, &ops, &paths);
+        if (r < 0) {
+                log_warning_errno(r, "Failed to parse DBus introspect XML, ignoring: %m");
+                return 0;
+        }
+
+        SET_FOREACH(path, paths) {
+                if (streq(path, "/org/freedesktop/ModemManager/Bearer"))
+                        continue;
+
+                (void) bearer_new_and_initialize(manager, path);
+        }
+
+        return 0;
+}
+
+int manager_enumerate_bearers(Manager *manager) {
+        int r;
+
+        assert(manager);
+        assert(sd_bus_is_ready(manager->bus) > 0);
+
+        r = sd_bus_call_method_async(
+                        manager->bus,
+                        NULL,
+                        "org.freedesktop.ModemManager1",
+                        "/org/freedesktop/ModemManager/Bearer",
+                        "org.freedesktop.DBus.Introspectable",
+                        "Introspect",
```

Comment:
hmm, doesn't mman implement ObjectManager? judging by https://www.freedesktop.org/software/ModemManager/api/1.0.0/ref-dbus-standard-interfaces-objectmanager.html it does.

I think that's a much better way to enumerate objects, if there's no explicit way to enumerate them.

Reply:
The only way to enumerate the **modem** objects is via the ObjectManager interface. Then, once you have the modem object, you can see the list of bearer objects associated to each modem in the "Bearers" property of the "Modem" interface in the modem object, see https://www.freedesktop.org/software/ModemManager/doc/latest/ModemManager/gdbus-org.freedesktop.ModemManager1.Modem.html#gdbus-property-org-freedesktop-ModemManager1-Modem.Bearers

Reply:
The usual way to work with this would be to get the initial list of modem objects and the added/removed events from the ObjectManager interface; and then monitor property updates in the "Bearers" property of the "Modem" object to get notified about bearer additions/removals, and then monitor the property updates of each of the bearers. It's quite some work without a helper library like libmm-glib, truth be told.

If you truly are only interested in monitoring the bearer objects, listening for property updates on `/org/freedesktop/ModemManager/Bearer` objects doesn't seem that bad idea.

Just beware, because there will be bearer objects reported as connected but without any data interface associated, e.g. the initial EPS bearer
```
aleksander@hestia:~$ mmcli -m a
...
  -----------------------------------
  3GPP EPS |    ue mode of operation: csps-2
           |     initial bearer path: /org/freedesktop/ModemManager1/Bearer/0
...

aleksander@hestia:~$ mmcli -b 0
  -------------------------
  General    |        path: /org/freedesktop/ModemManager1/Bearer/0
             |        type: default-attach
  -------------------------
  Status     |   connected: yes
             |   suspended: no
             | multiplexed: no
             |  ip timeout: 20
  -------------------------
  Properties |         apn: telefonica.es
             |     ip type: ipv4

```

You should ignore those bearers with type `default-attach` and without any interface reported.